src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -106,8 +106,8 @@ public NodeStateTable<T> AsCached()
 
         public T Single()
         {
-            Debug.Assert(this._states.Length == 1 && this._states[0].Count == 1);
-            return this._states[0].GetItem(0);
+            Debug.Assert((_states.Length == 1 || _states.Length == 2 && _states[0].IsRemoved) && this._states[^1].Count == 1);
```

Comment:
>_states.Length == 2 && _states[0].IsRemoved [](http://example.com/codeflow?start=49&length=43)

I don't think I understand this. If `Single()` supports two states where the first is marked `IsRemoved`, are there also valid cases with more than two states where all but the last are marked `IsRemoved`?

Reply:
Sort of, but not currently in the way the `NodeStateTable` is used. In the cases where nodes call `Single()` there should ever be one element, or one removed and a new one that replaces it. We could move the stronger assertion out to the callsites, and allow the state table to contain multiple removed elements, but given that nothing else uses it right now, it seems fine to leave it here.